File: 200_LeetCode_Questions/C++/D09P3_LeetCode_304.cpp
class NumMatrix {
private:
    vector<vector<int>> full;
    
public:
    NumMatrix(vector<vector<int>>& matrix) {
        if(matrix.size()==0) return;
        vector<int> t;
    
        int row = matrix.size(),col = matrix[0].size();
        
        
        for(int i=0;i<col+1;i++)
        {
            t.push_back(0);
        }
        full.push_back(t);
        
        for(int i=0;i<row;i++)
        {
            full.push_back(t);
        }
        cout<<"row="<<row;
        cout<<"col="<<col;
        cout<<full.size();
        for(int i=1;i<=row;i++)
        {
            for(int j=1;j<=col;j++)
            {
                full[i][j]=full[i-1][j]+full[i][j-1]+matrix[i-1][j-1]-full[i-1][j-1]; 
            }
        }
        
    }
    
    int sumRegion(int row1, int col1, int row2, int col2) {
        return full[row2+1][col2+1]-full[row1][col2+1]-full[row2+1][col1]+full[row1][col1];
        // return 0;
    }
};

/**
 * Your NumMatrix object will be instantiated and called as such:
 * NumMatrix* obj = new NumMatrix(matrix);
 * int param_1 = obj->sumRegion(row1,col1,row2,col2);
 */

File: 200_LeetCode_Questions/Java/D09P3_LeetCode_304.java
class Main {

  public static int[][] matrix = { 
      { 3, 0, 1, 4, 2 }, 
      { 5, 6, 3, 2, 1 }, 
      { 1, 2, 0, 1, 5 }, 
      { 4, 1, 0, 1, 7 },
      { 1, 0, 3, 0, 5 } 
  };

  public static void main(String args[]) {
    System.out.println(sumRegion(2, 1, 3, 4));
  }

  public static int sumRegion(int row1, int col1, int row2, int col2) {
    if (row2 < row1 || col2 < col1) {
      throw new ArithmeticException("Input row or col Err"); 
    }
    int sum = 0;
    for (int i = col1; i <= col2; i++) {
      for (int j = row1; j <= row2; j++) {
        sum += matrix[j][i];
      }
    }
    return sum;
  }
}